string_match_without_scan_or_regex.rb

```diff
@@ -1,36 +1,63 @@
-def match?(string, substring)
-  string_array = string.split("")
-  substring_array = substring.split("")
-  character_matches = []
-  substring_array.each_with_index do |substring_character, substring_index|
-    find_character_match(string_array, substring_array, substring_character, substring_index, character_matches)
+class String
+
+  def match?(substring)
+    if substring.class == String
+      #passing hash to collect matches, as well as the starting
+      #index 0 for the recursive call
+      matches = generate_matches(self, substring, {}, 0)
+      matches.has_value?(substring)
+    else
+      wrong_argument_type
+    end
   end
-  character_matches.join("") == substring ? true : false
-end
 
-def find_character_match(string_array, substring_array, substring_character, substring_index, character_matches)
-  if string_array.include?(substring_character)
-    if substring_index == 0
-      character_matches.push(substring_character)
-    elsif substring_index > 0
-      string_array.each_with_index do |string_character, string_index|
-
-        characters_do_match = string_character == substring_character
-        previous_string_index_not_negative = string_index - 1 > -1
-        previous_characters_do_match = string_array[string_index - 1] == substring_array[substring_index - 1]
-
-        if characters_do_match && previous_string_index_not_negative && previous_characters_do_match
-          character_matches.push(substring_character)
-          break
-        end
-      end
+  def match_count(substring)
+    if substring.class == String
```

Comment:
in a dynamically typed language, you generally don't need to/shouldn't be checking types.  Can you think of a way to write this method to return the appropriate value for several different data types?